programs/marinade-finance/src/instructions/management/redelegate.rs

```diff
@@ -0,0 +1,368 @@
+        let re_delegate_amount = if desired_redelegate_amount > max_redelegate_from_validator {
+            max_redelegate_from_validator
+        } else {
+            desired_redelegate_amount
+        };
+        // compute how much this particular account will have after split
+        let stake_account_after = stake
+            .last_update_delegated_lamports
+            .saturating_sub(re_delegate_amount);
+
+        // get dest validator from index
+        let mut dest_validator = self.state.validator_system.get(
+            &self.validator_list.data.as_ref().borrow(),
+            dest_validator_index,
+        )?;
+        // verify dest_validator_account matches
+        if self.dest_validator_account.key() != dest_validator.validator_account {
+            return err!(MarinadeError::IncorrectDestValidatorAccountOrIndex);
+        }
+
+        // compute dest validator target
+        let dest_validator_stake_target = self
+            .state
+            .validator_system
+            .validator_stake_target(&dest_validator, total_stake_target)?;
+        // if added stake will put the validator over target, reject
+        if dest_validator.active_balance + re_delegate_amount > dest_validator_stake_target {
+            msg!(
+                "Dest validator {} stake {} + amount {} is > target {}",
+                dest_validator.validator_account,
+                dest_validator.active_balance,
+                re_delegate_amount,
+                dest_validator_stake_target
+            );
+            return err!(MarinadeError::RedelegateOverTarget);
+        }
+
+        // select if we redelegate all or if we split first
+        // (do not leave less than min_stake in the account)
+        let (source_account, redelegate_amount_from_source_account) =
+            if stake_account_after < self.state.stake_system.min_stake {
+                // redelegate all if what will remain in the account is < min_stake
+                msg!("ReDelegate whole stake {}", stake.stake_account);
+
+                // Return back the rent reserve of unused split stake account
+                withdraw(
+                    CpiContext::new(
+                        self.stake_program.to_account_info(),
+                        Withdraw {
+                            stake: self.split_stake_account.to_account_info(),
+                            withdrawer: self.split_stake_account.to_account_info(),
+                            to: self.split_stake_rent_payer.to_account_info(),
+                            clock: self.clock.to_account_info(),
+                            stake_history: self.stake_history.to_account_info(),
+                        },
+                    ),
+                    self.split_stake_account.to_account_info().lamports(),
+                    None,
+                )?;
+
+                // mark as emergency_unstaking, the account will be cooling down
+                stake.is_emergency_unstaking = 1;
```

Comment:
Emergency unstaking state makes no difference for last_update_delegated_lamports = 0 as I remember. So looks like better to keep it in normal unstake mode to not surprise readers